packages/flutter/lib/src/material/navigation_bar.dart

```diff
@@ -0,0 +1,1169 @@
+  /// Builds a destination (icon + label) to use in a Material 3 [NavigationBar].
+  const _NavigationDestinationBuilder({
+    Key? key,
+    required this.buildIcon,
+    required this.buildLabel,
+    required this.label,
+    this.tooltip,
+  }) : super(key: key);
+
+  /// Builds the icon for an destination in a [NavigationBar].
+  ///
+  /// To animate between unselected and selected, build the icon based on
+  /// [NavigationDestinationInfo.selectedAnimation]. When the animation is 0,
+  /// the destination is unselected, when the animation is 1, the destination is
+  /// selected.
+  ///
+  /// The destination is considered selected as soon as the animation is
+  /// increasing or completed, and it is considered unselected as soon as the
+  /// animation is decreasing or dismissed.
+  final WidgetBuilder buildIcon;
+
+  /// Builds the label for an destination in a [NavigationBar].
+  ///
+  /// To animate between unselected and selected, build the icon based on
+  /// [NavigationDestinationInfo.selectedAnimation].  When the animation is
+  /// 0, the destination is unselected, when the animation is 1, the destination
+  /// is selected.
+  ///
+  /// The destination is considered selected as soon as the animation is
+  /// increasing or completed, and it is considered unselected as soon as the
+  /// animation is decreasing or dismissed.
+  final WidgetBuilder buildLabel;
+
+  /// The text value of what is in the label widget, this is required for
+  /// semantics so that screen readers and tooltips can read the proper label.
+  final String label;
+
+  /// The text to display in the tooltip for this [NavigationDestination], when
+  /// the user long presses the destination.
+  ///
+  /// If [tooltip] is an empty string, no tooltip will be used.
+  ///
+  /// Defaults to null, in which case the [label] text will be used.
+  final String? tooltip;
+
+  @override
+  Widget build(BuildContext context) {
+    final NavigationDestinationInfo info = NavigationDestinationInfo.of(context);
+    return _NavigationBarDestinationSemantics(
+      child: _NavigationBarDestinationTooltip(
+        message: tooltip ?? label,
+        child: InkWell(
+          highlightColor: Colors.transparent,
+          onTap: info.onTap,
+          child: Row(
+            children: <Widget>[
+              Expanded(
+                child: _NavigationBarDestinationLayout(
+                  icon: buildIcon(context),
+                  label: buildLabel(context),
+                ),
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+
+/// Inherited widget for passing data from the [NavigationBar] to the
+/// [NavigationBar.destinations] children widgets.
+///
+/// Useful for building navigation destinations using:
+/// `NavigationDestinationInfo.of(context)`.
+///
+/// See [NavigationDestination] for an example.
+class NavigationDestinationInfo extends InheritedWidget {
```

Comment:
If this class is going to remain public, we'll need a simple example that demonstrates its use. The source code for NavigationDestination isn't such an example.